HCore-Storage/Client/Impl/AwsStorageClientImpl.cs

```diff
@@ -0,0 +1,663 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Http.Headers;
+using System.Threading.Tasks;
+using Amazon.S3;
+using Amazon.S3.Model;
+using Amazon.S3.Transfer;
+using HCore.Storage.Exceptions;
+using HCore.Storage.Helpers;
+using HCore.Storage.Models;
+using HCore.Storage.Models.AwsStorage;
+using HCore.Web.Exceptions;
+using Newtonsoft.Json;
+
+namespace HCore.Storage.Client.Impl
+{
+    public class AwsStorageClientImpl : IStorageClient
+    {
+        private const int _defaultPageSize = 1000;
+
+        private readonly IDictionary<string, string> _connectionInfoByKey;
+
+        public AwsStorageClientImpl(string connectionString)
+        {
+            _connectionInfoByKey = AwsHelpers.GetConnectionInfoByKey(connectionString);
+        }
+
+        public async Task DownloadToStreamAsync(string containerName, string fileName, Stream stream)
+        {
+            using var amazonS3 = AwsHelpers.GetAmazonS3(_connectionInfoByKey);
+
+            using var getObjectResponse = await GetObjectResponseAsync(amazonS3, containerName, fileName, required: true).ConfigureAwait(false);
+
+            await getObjectResponse.ResponseStream.CopyToAsync(stream).ConfigureAwait(false);
+        }
+
+        private static async Task<GetObjectResponse> GetObjectResponseAsync(IAmazonS3 amazonS3, string containerName, string fileName, bool required)
+        {
+            try
+            {
+                return await amazonS3.GetObjectAsync(containerName, fileName).ConfigureAwait(false);
+            }
+            catch (AmazonS3Exception amazonS3Exception)
+            {
+                if (amazonS3Exception.StatusCode == HttpStatusCode.NotFound)
+                {
+                    if (required)
+                    {
+                        throw new ExternalServiceApiException(ExternalServiceApiException.CloudStorageFileNotFound, "The file was not found");
+                    }
+
+                    return null;
+                }
+
+                if (amazonS3Exception.StatusCode == HttpStatusCode.Forbidden || amazonS3Exception.StatusCode == HttpStatusCode.Unauthorized)
+                {
+                    throw new ExternalServiceApiException(ExternalServiceApiException.CloudStorageFileAccessDenied, "Access to the file was denied");
+                }
+
+                throw;
+            }
+        }
+
+        public async Task<Stream> OpenReadAsync(string containerName, string fileName)
+        {
+            using var amazonS3 = AwsHelpers.GetAmazonS3(_connectionInfoByKey);
+
+            var getObjectResponse = await GetObjectResponseAsync(amazonS3, containerName, fileName, required: true).ConfigureAwait(false);
+
+            // Please note that we cannot dispose getObjectResponse here and ResponseStream.CanSeek is false
```

Comment:
Is CanSeek true for Azure?

Reply:
For Azure this method returns truly seekable stream 
Internally they use `LazyLoadingReadOnlyStream<TProperties>` which inherits from `Stream`
Do you want me to try using it instead of my basic "S3SeekableStream"?

Reply:
You will not be able to use that one because its tied to the Azure SDK implementation.
The question is why we need the seeking.